### PR TITLE
 Search by name

### PR DESCRIPTION
**Pull request by dave0100**
Inital add of a fluent API for searches across a given BoxNavigation state.

See #26

### CODE REVIEW

**Comment by audax** [comment]
We should develop a mock for BoxNavigation objects to test this without having to set this up. Until now there wasn't really a need for it, since we tested the actual implementation. Now we have code like this that uses the interface only.

See #133

The code looks good so far, but I'll go to bed now ;)

**Comment by jan-schreib** [comment]
@audax No, it does not.
@dave0100 It seems you missed [this](https://github.com/Qabel/qabel-android/blob/master/CONTRIBUTING.md) completely.

File: qabelbox/src/androidTest/java/de/qabel/qabelbox/storage/SearchTest.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        Log.w(TAG, "NAV : "+nav);
+
+        debug(nav);
+
+    }
+
+    private void debug(BoxNavigation nav) throws Exception {
+
+        for(BoxFile file : nav.listFiles()) {
+            Log.w(TAG, "FILE: "+file.name);
```
Comment (audax) [inline qabelbox/src/androidTest/java/de/qabel/qabelbox/storage/SearchTest.java]: I think you meant to do a Log.d here

**Comment by dave0100** [comment]
@audax: Regarding using WARN-Level in Logging: I'll address that, too - for the time being I had to be able to filter the output easily.


In general: This pull request was not intended to actually be pulled yet, but for to be able to inspect early on (as suggested).


@Jan: Regarding the Contributing-Guides: I'll take a look at them in detail, thanks! I understand that you main concern is about the commits and pull request? If so, it would be nice to know how to make things right exactly. The initial commit for the feature should be in a single commit on an up-to-date fork of master, right? I'm not exactly sure how to reach this goal without making a new branch and pull reqeust.

And do you have additional concerns? This was in fact me first pull request as we use git without github @work

**Comment by jan-schreib** [comment]
@dave0100 Its more the formatting and indentation that is wrong. Except for not using a feature branch your request is fine.

**Comment by dave0100** [comment]
@jan-schreib That's a sane and fitting guideline (I especially like the part about the documentation. Too much and nobody bothers to read anything) - it was just that I didn't read until today and my defaults were way off. I hopefully fixed most of it and setting those defaults may prevent this in the future.

@audax The loglevel is debug now, too. Will work on the missing API calls now... can you tell me what you need in addition to consider the task done?

**Comment by jan-schreib** [comment]
@dave0100 No need to explain yourself to me, I'm just here to annoy you with our standards. ;) There are a lot of mixed indentations in the Android code at the moment. I will discuss that with @audax since  this is not the only pullrequest that has/had this issue.

**Comment by audax** [comment]
I think what is missing in the API is a mapping BoxObject -> Path, so that the search GUI can navigate to that folder. Also, we can only access BoxObjects if we know in which folder they are.

So in the initial collectAll step you could create a mapping of BoxObject -> Path and also allow it to be passed as an initial argument.

Sorry that I missed that earlier.

**Comment by dave0100** [comment]
Ah ok, I will add that!

**Comment by dave0100** [comment]
@audax FYI: I first tried to use BoxObject as the key for the Hashtable, but since equals and hashcode are not implemented this failed occasionally, and I used the path as the key instead and wrote a reverse lookup helper method which explicitly casts and calls the corresponding equals-methods.

Please let me know if that works for you and if you need anything else done. Refactoring in case you want to take another approch would be no problem, of course.

I also created two clone methods in my test class since clone() was not implemented, in case you want to reuse them or already have a helper method I can use instead.

**Comment by audax** [comment]
The latest commit switched to spaces from tabs which messes up the convenient diff view on github :tired_face: 
Whitespace, oh bane of my existence. Feel free to amend that commit and force push it.

Also, you can and are encouraged to just improve existing classes, like adding equals, hashcode and clone to BoxObject and it's subclasses. No need to work around stuff we can easily change.

**Comment by dave0100** [comment]
Just when I though I had configured it right O_o

I'll fix that, regarding the clone-methods and hashcode/equals I'm not sure if it would be the right approach to implement it on BoxObject. It would be possible to implement them, but the I'd have to cast to subclasses of all possible classes there inside of BoxObject. This will be error prone (especially if new types are added), and I don't see an easy, maintainable way. 

Like described here in Answer 1 there seems no sane solution:
http://stackoverflow.com/questions/13162188/java-equals-method-in-base-class-and-in-subclasses

Maybe if BoxObject would be an interface instead it would work (because the implementing classes methods would be used instead), but I don't think a change this big would be appropriate right now, considered the feature works and it's convenient to be able to create BoxObject in other places easily.

What do you think?

**Comment by dave0100** [comment]
@audax The line of thinking behind this is of course that name in BoxObject is not sufficient for equals, because multiple files/folders in different subclasses may have the same name.

**Comment by audax** [comment]
Sometimes, I don't like Java very much.

You convinced me about the methods in BoxObject, let's keep it like it is now. Code that wants to clone/compare a BoxObject usually does so for a very specific purpose and probably wouldn't use the generic methods anyway.

**Comment by dave0100** [comment]
@audax Yeah, I even still don't fully understand why I got errors looking up the exact same objects. The tabs issue is fixed now, seems like I may not use "auto-indent lines" even after switching my preferences to "use tabs"...

**Pull request merged**